internal/impl/snowflake/streaming/streaming.go

```diff
@@ -272,72 +271,132 @@ func (c *SnowflakeIngestionChannel) nextRequestID() string {
 // InsertStats holds some basic statistics about the InsertRows operation
 type InsertStats struct {
 	BuildTime            time.Duration
+	ConvertTime          time.Duration
+	SerializeTime        time.Duration
 	UploadTime           time.Duration
 	CompressedOutputSize int
 }
 
+type bdecPart struct {
+	unencryptedLen  int
+	parquetFile     []byte
+	parquetMetadata format.FileMetaData
+	stats           []*statsBuffer
+	convertTime     time.Duration
+	serializeTime   time.Duration
+}
+
+func (c *SnowflakeIngestionChannel) constructBdecPart(batch service.MessageBatch, metadata map[string]string) (bdecPart, error) {
+	wg := &errgroup.Group{}
+	type rowGroup struct {
+		rows  []parquet.Row
+		stats []*statsBuffer
+	}
+	rowGroups := []rowGroup{}
+	const maxRowGroupSize = 50_000
+	convertStart := time.Now()
+	for i := 0; i < len(batch); i += maxRowGroupSize {
+		end := min(maxRowGroupSize, len(batch[i:]))
+		j := len(rowGroups)
+		rowGroups = append(rowGroups, rowGroup{})
+		chunk := batch[i : i+end]
+		wg.Go(func() error {
+			rows, stats, err := constructRowGroup(chunk, c.schema, c.transformers)
+			rowGroups[j] = rowGroup{rows, stats}
+			return err
+		})
+	}
```

Comment:
I wonder if this combined with the default `max_in_flight` value (64) can lead to too much concurrency issues.

Reply:
Yeah I guess I haven't seen past parallelism issues. In theory this is purely compute bound work and I was not seeing full utilization without this (actually I don't see full utilization with this I need to parallelize compression too, but I have to switch to arrow for that).

I will add a limit to the err group and plumb that down as an advanced config option called build_parallelism.

Reply:
I haven't really done much benchmarking, so dunno how it might behave. I was just thinking that the scheduler might end up doing a bit of extra context switching if the number of goroutines is much bigger than the number of cores for CPU-intensive tasks, but maybe it's smart enough to not do that. I might toy around with an example when I get a bit of time.

Reply:
The runtime can preempt long running goroutines and context switch so limiting this is better. I will just surface this as a config

Reply:
Done - thanks for this suggestion!